cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/SystemDatasetRuntimeModule.java

```diff
@@ -88,7 +121,7 @@ protected void configure() {
                                                                             Names.named("defaultDatasetModules"));
 
         // NOTE: order is important due to dependencies between modules
-        mapBinder.addBinding("orderedTable-hbase").toInstance(new HBaseTableModule());
+        mapBinder.addBinding("orderedTable-hbase").toInstance(tableModule);
```

Comment:
My previous comment mean turning this binding into a Provider binding so that `CConfiguration` is injected, rather than passed to through `getDistributedModules()` method. Something like this:

```java
@Override
public Module getDistributedModules() {
  return new AbstractModule() {
    @Override
    protected void configure() {
      ...
      // NOTE: order is important due to dependencies between modules
      mapBinder.addBinding("orderedTable-hbase").toProvider(CoreTableModuleProvider.class).in(Scopes.SINGLETON);
      ...
    }
  };
}

private static final class CoreTableModuleProvider implements Provider<DatasetModule> {

  private final CConfiguration cConf;
  private final Injector injector;

  private CoreTableModuleProvider(CConfiguration cConf, Injector injector) {
    this.cConf = cConf;
    this.injector = injector;
  }

  @Override
  public DatasetModule get() {
    String tableModuleName = getDistributedTableModuleName();
    if (tableModuleName == null) {
      return new HBaseTableModule();
    }

    try {
      return (DatasetModule) Class.forName(tableModuleName).newInstance();
    } catch (ClassNotFoundException | ClassCastException | InstantiationException | IllegalAccessException ex) {
      throw new RuntimeException(ex);
    }
  }

  private String getDistributedTableModuleName() {
    String name = cConf.get(Constants.Dataset.Extensions.DISTMODE_TABLE);
    return name == null ? null : name.trim();
  }
}

```